invariance.md

```diff
@@ -0,0 +1,222 @@
+> If all data members can vary independently of each other, no invariant is possible.
+>
+> #### Example
+>
+> ```cpp
+>     struct Pair {  // the members can vary independently
+>         string name;
+>         int volume;
+>     };
+> ```
+>
+> but:
+>
+> ```cpp
+>     class Date {
+>     public:
+>         // validate that {yy, mm, dd} is a valid date and initialize
+>         Date(int yy, Month mm, char dd);
+>         // ...
+>     private:
+>         int y;
+>         Month m;
+>         char d;    // day
+>     };
+> ```
+>
+> #### Note
+>
+> If a class has any `private` data, a user cannot completely initialize an object without the use of a constructor.
+> Hence, the class definer will provide a constructor and must specify its meaning.
+> This effectively means the definer needs to define an invariant.
+>
+> **See also**:
+>
+> * [define a class with private data as `class`](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-class)
+> * [Prefer to place the interface first in a class](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rl-order)
+> * [minimize exposure of members](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-private)
+> * [Avoid `protected` data](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rh-protected)
+>
+> #### Enforcement
+>
+> Look for `struct`s with all data private and `class`es with public members.
+
+## But what about when my invariants or data change?
+
+One potential critique of these guidelines for choosing a struct versus a class is that public member functions provide an interface that isolates the private data members from the users of that object. If I have a class with a getter, I can change the type of that private data without modifying calls to the getter as long as the return type remains the same. If a struct is used and the data members are manipulated directly, any change to the types of those data members impacts the struct's consumer. Additionally, if a struct is used and we need to introduce an invariance for the private data (which would necessitate a switch from a struct to a class), then we'll also have to modify all the callers. Using a class is less likely to cause the callers of its member functions to change.
+
+I agree that this is definitely a tradeoff. I think a couple of things can serve to minimize the impact of these scenarios. First, [DMI](dmi.md) gives guidance that keeps the scope of data small, meaning that the scope of updates when objects change will also be small. Second, changing from setting a variable directly to calling a getter or setter when a new invariance is introduced is a minor refactor. Third, the placement of data is often just a best guess at first, so the location, type, etc. of data is likely to change over time anyway. The setter/getter interface is also fluid early in development. Modern IDEs significantly simplify these types of refactors, and the introduction of AI will only continue this trend.
```

Comment:
Not being familiar with DMI, I don't know whether it covers classes or structs, so I don't know in which context to frame the guidance. The "First, DMI..." sentence leads me to assume it's regarding structs. The "Second, ..." sentence throws me off because now we're talking about refactoring and I don't even know for sure from what to what yet. The "Third, ..." sentence contributes very little to my understanding; I can think about it and infer that refactoring is part of coding life. "The setter/getter" sentence feels like an after though that doesn't connect to the last sentence's "these types of refactors" maybe only because referring to the "setter/getter interface" as "fluid" doesn't clarify to me what you mean entirely.